In [1]:
import pandas as pd
import warnings
import numpy as np
import matplotlib.pyplot as plt
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_colwidth', 10000)

# SO Data

Our goal is to compare between the bachelor degree to master degree's salaries in stack overflow survey.
```
df['group1'] = np.where(df['FormalEducation'].str.contains('Master', case=False, na=False), 1, 0)

df['group2'] = np.where(df['FormalEducation'].str.contains('Bachelor', case=False, na=False), 1, 0)
```

In [2]:
df = pd.read_csv("outputs/so/clean_data.csv")
avg_group1 = np.mean(df.loc[df['group1']==1]['ConvertedCompYearly'])
avg_group2 = np.mean(df.loc[df['group2']==1]['ConvertedCompYearly'])
print(f"mean salary group1 = master degree is: {avg_group1}")
print(f"mean salary group2 = bachelor degree is: {avg_group2}")
print(f"salary mean diff is: {avg_group1-avg_group2}")

mean salary group1 = master degree is: 539667.8574363188
mean salary group2 = bachelor degree is: 635203.6378091428
salary mean diff is: -95535.78037282405


The negative diff means that in the survey master degrees got lower salaries.

Now we want to search if there are groups that have opposite relation (positive diff).

Lets look at all the groups that have at least 0.1<support and are defined with the 
attributes = ['Gender', 'Hobby', 'Student', 'LastNewJob', 'Exercise']


We will run the DivExplorer code:

In [3]:
df_step1 = pd.read_csv("outputs/so/interesting_subpopulations.csv")
df_step1

support  \
0   0.103109   
1   0.104814   
2   0.149631   
3   0.110792   
4   0.124284   
5   0.130448   
6   0.122106   
7   0.132018   
8   0.225468   
9   0.163022   
10  0.156977   
11  0.174674   
12  0.279826   
13  0.198889   
14  0.130060   
15  0.219203   
16  0.160692   
17  0.132998   
18  0.790810   
19  0.112396   
20  0.102822   
21  0.126969   
22  0.167396   
23  0.209190   
24  0.109374   
25  0.115301   
26  0.617808   
27  0.142691   
28  0.613114   
29  0.789307   
30  0.497813   
31  0.260845   
32  0.206994   
33  0.172765   
34  0.213749   
35  0.214087   
36  0.102687   
37  0.167818   
38  0.171499   
39  0.411338   
40  0.511677   
41  0.100339   
42  0.185211   
43  0.141998   
44  0.178017   
45  0.149834   
46  0.181631   
47  0.160371   
48  0.221635   
49  0.150239   
50  0.126699   
51  0.115757   
52  0.144920   
53  0.178541   
54  0.123069   
55  0.120603   
56  0.148432   
57  0.134079   
58  0.110590   
59  0.111147   
60  0.150965   
61  0.110826   
62  0.181850   
63  0.145595   
64  0.122427   

                                                                                         itemset  \
0                        frozenset({'Exercise=3 - 4 times per week', 'Hobby=Yes', 'Student=No'})   
1                       frozenset({'Gender=Male', 'Hobby=Yes', 'Exercise=3 - 4 times per week'})   
2                                                          frozenset({'Student=Yes, full-time'})   
3                      frozenset({'Gender=Male', 'Student=No', 'Exercise=3 - 4 times per week'})   
4                                      frozenset({'Exercise=3 - 4 times per week', 'Hobby=Yes'})   
5                                    frozenset({'Gender=Male', 'Exercise=3 - 4 times per week'})   
6                                             frozenset({'Hobby=Yes', 'Student=Yes, full-time'})   
7                                     frozenset({'Exercise=3 - 4 times per week', 'Student=No'})   
8                                    frozenset({'LastNewJob=Less than a year ago', 'Hobby=Yes'})   
9                     frozenset({'Gender=Male', 'LastNewJob=Less than a year ago', 'Hobby=Yes'})   
10                                                  frozenset({'Exercise=3 - 4 times per week'})   
11                     frozenset({'LastNewJob=Less than a year ago', 'Hobby=Yes', 'Student=No'})   
12                                                frozenset({'LastNewJob=Less than a year ago'})   
13                                 frozenset({'Gender=Male', 'LastNewJob=Less than a year ago'})   
14      frozenset({'Gender=Male', 'LastNewJob=Less than a year ago', 'Hobby=Yes', 'Student=No'})   
15                                  frozenset({'LastNewJob=Less than a year ago', 'Student=No'})   
16                   frozenset({'Gender=Male', 'LastNewJob=Less than a year ago', 'Student=No'})   
17                              frozenset({'LastNewJob=Between 2 and 4 years ago', 'Hobby=Yes'})   
18                                                                      frozenset({'Hobby=Yes'})   
19                frozenset({'LastNewJob=Between 2 and 4 years ago', 'Hobby=Yes', 'Student=No'})   
20               frozenset({'Gender=Male', 'LastNewJob=Between 2 and 4 years ago', 'Hobby=Yes'})   
21                            frozenset({'Gender=Male', 'LastNewJob=Between 2 and 4 years ago'})   
22                                           frozenset({'LastNewJob=Between 2 and 4 years ago'})   
23                                                                       frozenset({'Hobby=No'})   
24              frozenset({'Gender=Male', 'LastNewJob=Between 2 and 4 years ago', 'Student=No'})   
25                                                        frozenset({'Gender=Male', 'Hobby=No'})   
26                                                        frozenset({'Hobby=Yes', 'Student=No'})   
27                             frozenset({'LastNewJob=Between 2 and 4 years ago', 'Student=No'})   
28                                              

We can see in the first 16 rows there are different subpopulations with opposite relation!

The master(ConvertedCompYearly_group1 column) got higher salaries than the bachelor (ConvertedCompYearly_group2 column).

For example in the first row we see that the masters with fields: frozenset({'Exercise=3 - 4 times per week', 'Hobby=Yes', 'Student=No'}) averaged 1,052,558 but the bachelors averaged 758,811 -> diff of 293,746!

Now we want to see the reason, lets run the second step with the possible 

attributes = ['YearsCodingProf', 'Country','Age', 'RaceEthnicity', 'JobSatisfaction', 'HopeFiveYears', 'WakeTime', 'DevType']

In [4]:
df_step2 = pd.read_csv("outputs/so/subpopulations_and_treatments.csv")
df_step2

itemset  \
0                    frozenset({'Exercise=3 - 4 times per week', 'Hobby=Yes', 'Student=No'})   
1                   frozenset({'Gender=Male', 'Hobby=Yes', 'Exercise=3 - 4 times per week'})   
2                                                      frozenset({'Student=Yes, full-time'})   
3                  frozenset({'Gender=Male', 'Student=No', 'Exercise=3 - 4 times per week'})   
4                                  frozenset({'Exercise=3 - 4 times per week', 'Hobby=Yes'})   
5                                frozenset({'Gender=Male', 'Exercise=3 - 4 times per week'})   
6                                         frozenset({'Hobby=Yes', 'Student=Yes, full-time'})   
7                                 frozenset({'Exercise=3 - 4 times per week', 'Student=No'})   
8                                frozenset({'LastNewJob=Less than a year ago', 'Hobby=Yes'})   
9                 frozenset({'Gender=Male', 'LastNewJob=Less than a year ago', 'Hobby=Yes'})   
10                                              frozenset({'Exercise=3 - 4 times per week'})   
11                 frozenset({'LastNewJob=Less than a year ago', 'Hobby=Yes', 'Student=No'})   
12                                            frozenset({'LastNewJob=Less than a year ago'})   
13                             frozenset({'Gender=Male', 'LastNewJob=Less than a year ago'})   
14  frozenset({'Gender=Male', 'LastNewJob=Less than a year ago', 'Hobby=Yes', 'Student=No'})   
15                              frozenset({'LastNewJob=Less than a year ago', 'Student=No'})   
16               frozenset({'Gender=Male', 'LastNewJob=Less than a year ago', 'Student=No'})   
17                          frozenset({'LastNewJob=Between 2 and 4 years ago', 'Hobby=Yes'})   
18                                                                  frozenset({'Hobby=Yes'})   
19            frozenset({'LastNewJob=Between 2 and 4 years ago', 'Hobby=Yes', 'Student=No'})   
20           frozenset({'Gender=Male', 'LastNewJob=Between 2 and 4 years ago', 'Hobby=Yes'})   
21                                       frozenset({'LastNewJob=Between 2 and 4 years ago'})   
22                                                                   frozenset({'Hobby=No'})   
23                                                    frozenset({'Gender=Male', 'Hobby=No'})   
24                                                    frozenset({'Hobby=Yes', 'Student=No'})   
25                         frozenset({'LastNewJob=Between 2 and 4 years ago', 'Student=No'})   
26                                                                frozenset({'Gender=Male'})   
27                                                                 frozenset({'Student=No'})   
28                                                   frozenset({'Gender=Male', 'Hobby=Yes'})   
29                                        frozenset({"Exercise=I don't typically exercise"})   
30                           frozenset({"Exercise=I don't typically exercise", 'Hobby=Yes'})   
31            frozenset({'Gender=Male', "Exercise=I don't typically exercise", 'Hobby=Yes'})   
32                         frozenset({'Gender=Male', "Exercise=I don't typically exercise"})   
33                                         frozenset({'Exercise=Daily or almost every day'})   
34             frozenset({"Exercise=I don't typically exercise", 'Hobby=Yes', 'Student=No'})   
35                                                     frozenset({'Hobby=No', 'Student=No'})   
36                                     frozenset({'Gender=Male', 'Hobby=Yes', 'Student=No'})   
37                                                  frozenset({'Gender=Male', 'Student=No'})   
38                                      frozenset({'Gender=Male', 'Hobby=No', 'Student=No'})   
39                                       frozenset({'LastNewJob=Between 1 and 2 years ago'})   
40                         frozenset({'LastNewJob=Between 1 and 2 years ago', 'Student=No'})   
41                               frozenset({'Gender=M

We can see that there are different treatments for differenct itemsets.

In the itemset mentioned above (first row here as well) we got the explanation: Country=United States & RaceEthnicity=WhiteorofEuropeandescent

Lets check if it aligns with the data:

In [5]:
df_subpopulation = df.loc[(df["Exercise"]=="3 - 4 times per week") & (df["Hobby"]=="Yes") & (df["Student"]=="No")]
avg_group1 = np.mean(df_subpopulation.loc[df['group1']==1]['ConvertedCompYearly'])
avg_group2 = np.mean(df_subpopulation.loc[df['group2']==1]['ConvertedCompYearly'])
print(f"mean salary group1 = master degree is: {avg_group1}")
print(f"mean salary group2 = bachelor degree is: {avg_group2}")
print(f"salary mean diff is: {avg_group1-avg_group2}")

mean salary group1 = master degree is: 1052557.916747338
mean salary group2 = bachelor degree is: 758811.8455445544
salary mean diff is: 293746.0712027835


In [6]:
df_treated = df_subpopulation.loc[(df_subpopulation["Country"]=="United States") & (df_subpopulation["RaceEthnicity_WhiteorofEuropeandescent"]==1.0)]
avg_group1 = np.mean(df_treated.loc[df_treated['group1']==1]['ConvertedCompYearly'])
avg_group2 = np.mean(df_treated.loc[df_treated['group2']==1]['ConvertedCompYearly'])
print(f"mean salary group1 = master degree is: {avg_group1}")
print(f"mean salary group2 = bachelor degree is: {avg_group2}")
print(f"salary mean diff is: {avg_group1-avg_group2}")

mean salary group1 = master degree is: 2472876.4872521246
mean salary group2 = bachelor degree is: 1154974.7530864198
salary mean diff is: 1317901.7341657048


The actual results show that for this treatment the salary of the masters group had a greater increase that does not compare to the increase of the bachelors group salary, the difference was even bigger!

For the final results let pick the top K=5 facts from all the data:

In [7]:
df_step3 = pd.read_csv("outputs/so/find_k/5_0.5_9e-05.csv")
df_step3

itemset  \
0        {'Hobby': 'Yes', 'Exercise': '3 - 4 times per week', 'Student': 'No'}   
1                                                {'Student': 'Yes, full-time'}   
2  {'Hobby': 'Yes', 'Exercise': "I don't typically exercise", 'Student': 'No'}   
3                                             {'Hobby': 'No', 'Student': 'No'}   
4        {'Hobby': 'Yes', 'Exercise': '1 - 2 times per week', 'Student': 'No'}   

                                                                                                                                                                                                                                                                                   treatment  \
0  [{'att': 'Country', 'value': <function parse_treatment.<locals>.<lambda> at 0x00000198B5E98940>, 'val_specified': 'United States'}, {'att': 'RaceEthnicity_WhiteorofEuropeandescent', 'value': <function parse_treatment.<locals>.<lambda> at 0x00000198B5E988B0>, 'val_specified': 1.0}]   
1                     [{'att': 'Country', 'value': <function parse_treatment.<locals>.<lambda> at 0x00000198B5DE21F0>, 'val_specified': 'United States'}, {'att': 'Age', 'value': <function parse_treatment.<locals>.<lambda> at 0x00000198B5E98700>, 'val_specified': '25 - 34 years old'}]   
2                                                                                                                                     [{'att': 'RaceEthnicity_WhiteorofEuropeandescent', 'value': <function parse_treatment.<locals>.<lambda> at 0x00000198B5C9D8B0>, 'val_specified': 1.0}]   
3                                                                                                                                                        [{'att': 'Country', 'value': <function parse_treatment.<locals>.<lambda> at 0x00000198B5C9D5E0>, 'val_specified': 'United States'}]   
4                                                                                                                                                        [{'att': 'Country', 'value': <function parse_treatment.<locals>.<lambda> at 0x00000198B5C8B1F0>, 'val_specified': 'United States'}]   

           ate1          ate2        iscore  size_itemset  size_group1  \
0  1.713005e+06  5.832719e+05  1.129733e+06          6106         2066   
1  3.224831e+06  1.265303e+06  1.959529e+06          8861         2460   
2  2.381899e+05  5.677356e+05  3.295457e+05          9938         3240   
3  9.116676e+05  1.214402e+06  3.027349e+05         10156         3450   
4  6.784188e+05  1.640943e+06  9.625241e+05          8622         3224   

   size_group2   support  ni_score  utility           std     diff_means  
0         4040  0.103109       1.0      1.0  6.780992e+06  293746.071203  
1         6401  0.149631       1.0      1.0  3.557639e+06  208693.353272  
2         6698  0.167818       1.0      1.0  5.533553e+06 -222233.985090  
3         6706  0.171499       1.0      1.0  6.335351e+06 -222674.287851  
4         5398  0.145595       1.0      1.0  7.528571e+06 -490583.491174

# MEPS Data

Our goal is to compare between the young people to the old people if they are exercise (at least 3 times at week) or not.

Value 1 = "yes"

Value 2 = "no"

```
df['group1'] = df['AGE'].apply(lambda x: 1 if x <= 30 else 0)

df['group2'] = df['AGE'].apply(lambda x: 1 if x >= 60 else 0)
```

In [8]:
df = pd.read_csv("outputs/meps/clean_data.csv")
avg_group1 = np.mean(df.loc[df['group1']==1]['Exercise'])
avg_group2 = np.mean(df.loc[df['group2']==1]['Exercise'])
print(f"mean exercise group1 = young is: {avg_group1}")
print(f"mean exercise group2 = old is: {avg_group2}")
print(f"mean diff is: {avg_group1-avg_group2}")

mean exercise group1 = young is: 1.4385964912280702
mean exercise group2 = old is: 1.5389841475250727
mean diff is: -0.10038765629700253


The negative diff means that the younger do more exercise than the older.

Now we want to search if there are groups that have opposite relation (positive diff).

Lets look at all the groups that have at least 0.1<support and are defined with the 
attributes = ['Married', 'IsHadStroke', 'DoesDoctorRecommendExercise', 'IsWorking', 'CurrentlySmoke']


We will run the DivExplorer code:

In [9]:
df_step1 = pd.read_csv("outputs/meps/interesting_subpopulations.csv")
df_step1

support  \
0   0.312721   
1   0.315612   
2   0.319949   
3   0.339544   
4   0.344523   
5   0.522647   
6   0.366849   
7   0.585288   
8   0.542724   
9   0.407485   
10  0.413588   
11  0.368615   
12  0.457437   
13  0.464664   
14  0.608577   
15  0.402666   
16  0.418407   
17  0.371667   
18  0.850626   
19  0.441857   
20  0.308384   
21  0.422904   
22  0.309830   
23  0.442178   
24  0.327979   
25  0.946033   
26  0.896884   
27  0.360745   
28  0.487954   
29  0.354160   
30  0.391423   

                                                                                  itemset  \
0      frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0', 'IsWorking=1.0'})   
1                         frozenset({'DoesDoctorRecommendExercise=2.0', 'IsWorking=1.0'})   
2                       frozenset({'IsHadStroke=2.0', 'CurrentlySmoke=0', 'Married=1.0'})   
3                                          frozenset({'CurrentlySmoke=0', 'Married=1.0'})   
4                                           frozenset({'IsHadStroke=2.0', 'Married=1.0'})   
5   frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0', 'CurrentlySmoke=0'})   
6                                                              frozenset({'Married=1.0'})   
7                       frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0'})   
8                      frozenset({'DoesDoctorRecommendExercise=2.0', 'CurrentlySmoke=0'})   
9                     frozenset({'IsHadStroke=2.0', 'CurrentlySmoke=0', 'IsWorking=1.0'})   
10                                       frozenset({'CurrentlySmoke=0', 'IsWorking=1.0'})   
11                      frozenset({'IsHadStroke=2.0', 'CurrentlySmoke=0', 'Married=5.0'})   
12                                        frozenset({'IsHadStroke=2.0', 'IsWorking=1.0'})   
13                                                           frozenset({'IsWorking=1.0'})   
14                                         frozenset({'DoesDoctorRecommendExercise=2.0'})   
15                    frozenset({'CurrentlySmoke=0', 'IsHadStroke=2.0', 'IsWorking=4.0'})   
16                                          frozenset({'IsHadStroke=2.0', 'Married=5.0'})   
17                                         frozenset({'CurrentlySmoke=0', 'Married=5.0'})   
18                                     frozenset({'IsHadStroke=2.0', 'CurrentlySmoke=0'})   
19                                        frozenset({'IsHadStroke=2.0', 'IsWorking=4.0'})   
20       frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0', 'Married=5.0'})   
21                                                             frozenset({'Married=5.0'})   
22                          frozenset({'DoesDoctorRecommendExercise=2.0', 'Married=5.0'})   
23                                       frozenset({'CurrentlySmoke=0', 'IsWorking=4.0'})   
24  frozenset({'IsHadStroke=2.0', 'CurrentlySmoke=0', 'DoesDoctorRecommendExercise=1.0'})   
25                                                         frozenset({'IsHadStroke=2.0'})   
26                                                        frozenset({'CurrentlySmoke=0'})   
27                      frozenset({'IsHadStroke=2.0', 'DoesDoctorRecommendExercise=1.0'})   
28                                                           frozenset({'IsWorking=4.0'})   
29                     frozenset({'CurrentlySmoke=0', 'DoesDoctorRecommendExercise=1.0'})   
30                                         frozenset({'DoesDoctorRecommendExercise=1.0'})   

    GROUP_1  GROUP_2  Exercise  Exercise_group1  Exercise_group2  \
0    1549.0    398.0  1.381613         1.395739         1.326633   
1    1552.0    413.0  1.381679         1.394974         1.331719   
2     571.0   1421.0  1.495482         1.509632         1.489796   
3     574.0   1540.0  1.506149         1.508711         1.505195   
4     624.0   1521.0  1.499767         1.501603         1.499014   
5    2089.0   1165.0  1.430854         1.427956         1.436052   
6     627.0   1657.0  1.510946         

We can see in the first 5 rows that there are different subpopulations with opposite relation!

The older(Exercise_group1 column) got higher exercise average than the older (Exercise_group2 column)-> meaning they do less sport.

For example in the first row we will see that the yound group with fields: frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0', 'IsWorking=1.0'}) averaged 1.395739 but the old group averaged 1.326633	 -> diff of 0.069106!

Now we want to see the reason, lets run the second step with the posible

attributes = ['Region', 'Race', 'Sex', 'Education','IsHadHeartAttack','IsDiagnosedAsthma', 'IsBornInUSA',LongSinceLastFluVaccination', 'TakesAspirinFrequently','WearsSeatBelt', 'HoldHealthInsurance']

In [10]:
df_step2 = pd.read_csv("outputs/meps/subpopulations_and_treatments.csv")
df_step2

itemset  \
0      frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0', 'IsWorking=1.0'})   
1                         frozenset({'DoesDoctorRecommendExercise=2.0', 'IsWorking=1.0'})   
2                                          frozenset({'CurrentlySmoke=0', 'Married=1.0'})   
3                                           frozenset({'IsHadStroke=2.0', 'Married=1.0'})   
4   frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0', 'CurrentlySmoke=0'})   
5                                                              frozenset({'Married=1.0'})   
6                       frozenset({'DoesDoctorRecommendExercise=2.0', 'IsHadStroke=2.0'})   
7                      frozenset({'DoesDoctorRecommendExercise=2.0', 'CurrentlySmoke=0'})   
8                     frozenset({'IsHadStroke=2.0', 'CurrentlySmoke=0', 'IsWorking=1.0'})   
9                                        frozenset({'CurrentlySmoke=0', 'IsWorking=1.0'})   
10                                        frozenset({'IsHadStroke=2.0', 'IsWorking=1.0'})   
11                                                           frozenset({'IsWorking=1.0'})   
12                                         frozenset({'DoesDoctorRecommendExercise=2.0'})   
13                    frozenset({'CurrentlySmoke=0', 'IsHadStroke=2.0', 'IsWorking=4.0'})   
14                                     frozenset({'IsHadStroke=2.0', 'CurrentlySmoke=0'})   
15                                        frozenset({'IsHadStroke=2.0', 'IsWorking=4.0'})   
16                          frozenset({'DoesDoctorRecommendExercise=2.0', 'Married=5.0'})   
17                                       frozenset({'CurrentlySmoke=0', 'IsWorking=4.0'})   
18                                                         frozenset({'IsHadStroke=2.0'})   
19                                                        frozenset({'CurrentlySmoke=0'})   
20                                                           frozenset({'IsWorking=4.0'})   

               treatment  
0        IsBornInUSA_1.0  
1        IsBornInUSA_1.0  
2                 Race_4  
3          Education_5.0  
4   Race_3+Education_1.0  
5          Education_3.0  
6          Education_1.0  
7                 Race_3  
8        IsBornInUSA_2.0  
9        IsBornInUSA_1.0  
10       IsBornInUSA_1.0  
11       IsBornInUSA_1.0  
12         Education_1.0  
13       IsBornInUSA_1.0  
14                 Sex_1  
15       IsBornInUSA_1.0  
16         Education_1.0  
17       IsBornInUSA_1.0  
18                 Sex_1  
19                 Sex_1  
20       IsBornInUSA_1.0

We can see that there are different treatments for differenct itemsets.

In the itemset mentioned above (first row here as well) we got the explanation: IsBornInUSA=1

Lets check if it aligns with the data:

In [11]:
df_subpopulation = df.loc[(df["DoesDoctorRecommendExercise"]==2.0) & (df["IsHadStroke"]==2.0) & (df["IsWorking"]==1.0)]
avg_group1 = np.mean(df_subpopulation.loc[df['group1']==1]['Exercise'])
avg_group2 = np.mean(df_subpopulation.loc[df['group2']==1]['Exercise'])
print(f"mean exercise group1 = young is: {avg_group1}")
print(f"mean exercise group2 = old is: {avg_group2}")
print(f"mean diff is: {avg_group1-avg_group2}")

mean exercise group1 = young is: 1.395739186571982
mean exercise group2 = old is: 1.3266331658291457
mean diff is: 0.06910602074283623


In [12]:
df_treated = df_subpopulation.loc[df_subpopulation["IsBornInUSA"]==1.0]
avg_group1 = np.mean(df_treated.loc[df_treated['group1']==1]['Exercise'])
avg_group2 = np.mean(df_treated.loc[df_treated['group2']==1]['Exercise'])
print(f"mean exercise group1 = young is: {avg_group1}")
print(f"mean exercise group2 = old is: {avg_group2}")
print(f"mean diff is: {avg_group1-avg_group2}")

mean exercise group1 = young is: 1.3676703645007924
mean exercise group2 = old is: 1.288888888888889
mean diff is: 0.0787814756119034


The actual results show that for this treatment the young group didn't have much of a change compared to the old group which decreased to a lower difference, the difference got even bigger!

For the final results let pick the top K=5 facts from all the data:

In [13]:
df_step3 = pd.read_csv("outputs/meps/find_k/5_0.5_9e-05.csv")
df_step3

itemset  \
0  {'DoesDoctorRecommendExercise': 2.0, 'IsHadStroke': 2.0, 'CurrentlySmoke': 0.0}   
1                                          {'CurrentlySmoke': 0.0, 'Married': 1.0}   
2                             {'DoesDoctorRecommendExercise': 2.0, 'Married': 5.0}   
3                    {'IsWorking': 4.0, 'IsHadStroke': 2.0, 'CurrentlySmoke': 0.0}   
4       {'DoesDoctorRecommendExercise': 2.0, 'IsHadStroke': 2.0, 'IsWorking': 1.0}   

                                                                                                                                                                                                                                       treatment  \
0  [{'att': 'Race', 'value': <function parse_treatment.<locals>.<lambda> at 0x0000026E0CDDB790>, 'val_specified': 3.0}, {'att': 'Education', 'value': <function parse_treatment.<locals>.<lambda> at 0x0000026E0CDDBEE0>, 'val_specified': 1.0}]   
1                                                                                                                           [{'att': 'Race', 'value': <function parse_treatment.<locals>.<lambda> at 0x0000026E0CDDB5E0>, 'val_specified': 4.0}]   
2                                                                                                                      [{'att': 'Education', 'value': <function parse_treatment.<locals>.<lambda> at 0x0000026E0CE68670>, 'val_specified': 1.0}]   
3                                                                                                                    [{'att': 'IsBornInUSA', 'value': <function parse_treatment.<locals>.<lambda> at 0x0000026E0CE689D0>, 'val_specified': 1.0}]   
4                                                                                                                    [{'att': 'IsBornInUSA', 'value': <function parse_treatment.<locals>.<lambda> at 0x0000026E0CDDBB80>, 'val_specified': 1.0}]   

       ate1      ate2    iscore  size_itemset  size_group1  size_group2  \
0  0.273353  0.565891  0.292539          3254         2089         1165   
1  0.173622 -0.081288  0.254910          2114          574         1540   
2  0.064612  0.201058  0.136446          1929         1825          104   
3 -0.138174  0.058542  0.196716          2507          770         1737   
4 -0.151493 -0.180991  0.029497          1947         1549          398   

    support  ni_score   utility       std  diff_means  
0  0.522647  0.000292  0.000292  0.495196   -0.008096  
1  0.339544  0.000255  0.000255  0.499962    0.003516  
2  0.309830  0.000136  0.000136  0.491749   -0.075290  
3  0.402666  0.000197  0.000197  0.499436   -0.054873  
4  0.312721  0.000029  0.000029  0.485782    0.069106